src/crypto/equihash.cpp

```diff
@@ -19,183 +19,210 @@
-        throw invalid_params();
-    }
-    if ((n/(k+1)) % 8 != 0) {
-        std::cerr << "Parameters must satisfy n/(k+1) = 0 mod 8\n";
-        throw invalid_params();
-    }
-}
-
-int Equihash::InitialiseState(eh_HashState& base_state)
+template<unsigned int N, unsigned int K>
+int Equihash<N,K>::InitialiseState(eh_HashState& base_state)
 {
+    unsigned int n = N;
+    unsigned int k = K;
     unsigned char personalization[crypto_generichash_blake2b_PERSONALBYTES] = {};
     memcpy(personalization, "ZcashPOW", 8);
     memcpy(personalization+8,  &n, 4);
     memcpy(personalization+12, &k, 4);
     return crypto_generichash_blake2b_init_salt_personal(&base_state,
                                                          NULL, 0, // No key.
-                                                         n/8,
+                                                         N/8,
                                                          NULL,    // No salt.
                                                          personalization);
 }
 
-StepRow::StepRow(unsigned int n, const eh_HashState& base_state, eh_index i) :
-        hash {new unsigned char[n/8]},
-        len {n/8},
-        indices {i}
+void EhIndexToArray(const eh_index i, unsigned char* array)
 {
-    eh_HashState state;
-    state = base_state;
-    crypto_generichash_blake2b_update(&state, (unsigned char*) &i, sizeof(eh_index));
-    crypto_generichash_blake2b_final(&state, hash, n/8);
+    assert(sizeof(eh_index) == 4);
+    array[0] = (i >> 24) & 0xFF;
+    array[1] = (i >> 16) & 0xFF;
+    array[2] = (i >>  8) & 0xFF;
+    array[3] =  i        & 0xFF;
+}
 
-    assert(indices.size() == 1);
+eh_index ArrayToEhIndex(const unsigned char* array)
+{
+    assert(sizeof(eh_index) == 4);
+    eh_index ret {array[0]};
+    ret <<= 8;
+    ret |= array[1];
+    ret <<= 8;
+    ret |= array[2];
+    ret <<= 8;
+    ret |= array[3];
+    return ret;
 }
 
-StepRow::~StepRow()
+eh_trunc TruncateIndex(const eh_index i, const unsigned int ilen)
 {
-    delete[] hash;
+    // Truncate to 8 bits
+    assert(sizeof(eh_trunc) == 1);
+    return (i >> (ilen - 8)) & 0xff;
 }
 
-StepRow::StepRow(const StepRow& a) :
-        hash {new unsigned char[a.len]},
-        len {a.len},
-        indices(a.indices)
+eh_index UntruncateIndex(const eh_trunc t, const eh_index r, const unsigned int ilen)
 {
-    for (int i = 0; i < len; i++)
-        hash[i] = a.hash[i];
+    eh_index i{t};
+    return (i << (ilen - 8)) | r;
 }
 
-StepRow& StepRow::operator=(const StepRow& a)
+template<size_t WIDTH>
+StepRow<WIDTH>::StepRow(unsigned int n, const eh_HashState& base_state, eh_index i)
 {
-    unsigned char* p = new unsigned char[a.len];
-    for (int i = 0; i < a.len; i++)
-        p[i] = a.hash[i];
-    delete[] hash;
-    hash = p;
-    len = a.len;
-    indices = a.indices;
-    return *this;
+    eh_HashState state;
+    state = base_state;
+    crypto_generichash_blake2b_update(&state, (unsigned char*) &i, sizeof(eh_index));
```

Comment:
This is native endian-dependent. Do we have a ticket to fix that? I thought we did but I can't find it.